File: SmartGS_server/routers/order.py
from fastapi import APIRouter, Depends, HTTPException
from models.order import Order
from database import carts, orders, products
from routers.cart import get_current_user
from datetime import datetime
from bson import ObjectId

router = APIRouter(prefix="/order", tags=["Order"])

@router.post("/checkout")
def checkout(order: Order, payment_intent_id: str, username: str = Depends(get_current_user)):
    if not payment_intent_id:
        raise HTTPException(status_code=400, detail="Payment intent ID is required")
    
    enriched_items = []
    total_amount = 0.0

    for item in order.items:
        product = products.find_one({"id": item.product_id})
        if not product:
            raise HTTPException(status_code=404, detail=f"Product ID {item.product_id} not found")

        enriched_item = {
            "product_id": item.product_id,
            "product_name": item.product_name or product["name"],
            "quantity": item.quantity,
            "price": item.price or product["price"],
            "image": product.get("image", "no_image.png")
        }

        enriched_items.append(enriched_item)
        total_amount += enriched_item["price"] * item.quantity

    order_doc = {
        "username": username,
        "items": enriched_items,
        "total_amount": total_amount,
        "status": "paid",  
        "payment_intent_id": payment_intent_id,
        "timestamp": datetime.utcnow()
    }

    result = orders.insert_one(order_doc)
    order_id = str(result.inserted_id)

    carts.delete_many({"username": username})

    return {
        "message": "Order placed successfully",
        "order_id": order_id,
        "username": username,
        "total_amount": total_amount,
        "status": "paid",
        "items": enriched_items
    }

@router.get("/history")
def get_order_history(username: str = Depends(get_current_user)):
    try:
        user_orders = list(
            orders.find(
                {"username": username},
                {"_id": 1, "items": 1, "total_amount": 1, "status": 1, "timestamp": 1}
            ).sort("timestamp", -1)
        )
        
        for order in user_orders:
            order["order_id"] = str(order.pop("_id"))

        return {
            "username": username,
            "orders": user_orders,
            "count": len(user_orders)
        }

    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to load orders: {str(e)}")

File: SmartGS_client/lib/models/order.dart
class OrderItem {
  final int productId;
  final String productName;
  final int quantity;
  final double price;
  final String image;

  OrderItem({
    required this.productId,
    required this.productName,
    required this.quantity,
    required this.price,
    required this.image,
  });

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    return OrderItem(
      productId: json['product_id'],
      productName: json['product_name'],
      quantity: json['quantity'],
      price: (json['price'] as num).toDouble(),
      image: json['image'] ?? 'no_image.png',
    );
  }

  Map<String, dynamic> toJson() {
    return {'product_id': productId, 'quantity': quantity};
  }
}

class Order {
  final String orderId;
  final String username;
  final List<OrderItem> items;
  final double totalAmount;
  final String status;
  final DateTime timestamp;

  Order({
    required this.orderId,
    required this.username,
    required this.items,
    required this.totalAmount,
    required this.status,
    required this.timestamp,
  });

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      orderId: json['order_id'],
      username: json['username'] ?? '',
      items: (json['items'] as List)
          .map((item) => OrderItem.fromJson(item))
          .toList(),
      totalAmount: (json['total_amount'] as num).toDouble(),
      status: json['status'],
      timestamp: DateTime.parse(json['timestamp']),
    );
  }
}
